// Copyright 2015 The Swarming Authors. All rights reserved.
// Use of this source code is governed by the Apache v2.0 license that can be
// found in the LICENSE file.

// Schemas for project configs.

syntax = "proto2";

package buildbucket;

// A single access control rule.
message Acl {
  enum Role {
    // Can do read-only operations, such as search for builds.
    READER = 0;
    // Same as READER + can schedule and cancel builds.
    SCHEDULER = 1;
    // Can do all write operations.
    WRITER = 2;
  }
  // Role denotes a list of actions that an identity can perform.
  optional Role role = 1;
  // Name of the group defined in the auth service.
  optional string group = 2;
  // An email address or a full identity string "kind:name". See auth service
  // on kinds of identities. Anonymous users are "anonymous:anonymous".
  // Either identity or group must be present, not both.
  optional string identity = 3;
}

// Configuration of buildbucket-swarming integration for one bucket.
message Swarming {
  message Recipe {
    // Name of the recipe to run.
    optional string name = 1;
  }

  message Builder {
    // Name of the builder
    optional string name = 1;
    // Will be become to swarming task tags.
    // Each tag will end up in "swarming_tag" buildbucket tag, for example
    // "swarming_tag:builder:release"
    repeated string swarming_tags = 2;
    // Specifies that a recipe to run.
    optional Recipe recipe = 3;
    // Swarming task priority.
    optional int32 priority = 4;
  }

  // Hostname of the swarming instance, e.g. "chromium-swarm.appspot.com".
  optional string hostname = 1;
  // Used to generate a URL for Build, may contain parameters
  // {swarming_hostname}, {task_id}, {bucket} and {builder}. Defaults to:
  // https://{swarming_hostname}/user/task/{task_id}
  optional string url_format = 2;

  // Will be put to all swarming tasks.
  repeated string common_swarming_tags = 3;

  // Configuration for each builder.
  // Swarming tasks are created only for builds for builders that are not
  // explicitly specified.
  repeated Builder builders = 4;
}

// Defines one bucket in buildbucket.cfg
message Bucket {
  // Name of the bucket. Names are unique within one instance of buildbucket.
  // If another project already uses this name, a config will be rejected.
  // Name reservation is first-come first-serve.
  optional string name = 1;
  // List of access control rules for the bucket.
  repeated Acl acls = 2;
  // Buildbucket-swarming integration.
  optional Swarming swarming = 3;
}

// Schema of buildbucket.cfg file, a project config.
message BuildbucketCfg {
  // All buckets defined for this project.
  repeated Bucket buckets = 1;
}